### ISSUE TITLE
Universal admin interface

### ISSUE THREAD

**Issue by HelloZeroNet** [HelloZeroNet/ZeroNet#354]
We need an universal admin interface that allow to manage data stored on sites.

Features needed:
 - Query/Modify/Delete ZeroTalk and ZeroBlog (and other sites) database without any source code modifications. (using only configurationfiles)
 - Query/modify users limit using it

- [ ] Create sample configuration file for ZeroTalk and ZeroBlog (xml?, json?, js?, yaml?)
- [ ] Layout mockups
- [ ] ...Coding

This would be huge benefit for all ZeroNet sites, so it would be really nice if someone would be able to take this task, so I will be able to focus on core modifications.

**Comment by HelloZeroNet** [comment]
Example jsx config file for ZeroTalk:

```jsx
<Site title="ZeroTalk">

 <Page name="topic" title="Topics" orderby="added DESC">
  <user name="json_id" title="Username" disabled save="no"/>
  <input name="added" formatter={ Date.toDate } deformatter={ Date.fromDate }/>
  <input name="title"/>
  <textarea name="body"/>
  <select name="type" values={ ["", "group"] }/>
  <select name="parent_topic_uri" title="Parent topic" values={ Topics.listUris }/>
 </Page>

 <Page name="comment" title="Comments" orderby="added DESC" parent="topic">
  <user name="json_id" title="Username" disabled save="no"/>
  <input name="added" formatter={ Date.toDate } deformatter={ Date.fromDate }/>
  <textarea name="body"/>
  <select name="topic_uri" title="Topic" values={ Topics.listUris }/>
 </Page>

</Site>
```

Based on simple config this tool would able to generate an admin interface that allows the site owner to modify/delete/move topics and comments.

**Comment by icf20** [comment]
maybe you want to consider the a new option; preserve zeromail contact content across al hosting site

**Comment by alxbob** [comment]
Great option! Lets to it!

**Comment by vlad20012** [comment]
I'm the only one who does not understand why we need a "universal" interface, not just admin/user page on ZeroTalk/ZeroBlog?

**Comment by HelloZeroNet** [comment]
Because then you don't have to re-write the same functions for every site you make

**Comment by crazysoldier** [comment]
wow can someone give me the code for an admin interface on my own page?

**Comment by obv-mikhail** [comment]
So a content management system like WordPress, but for ZeroNet sites?

**Comment by obv-mikhail** [comment]
Is this supposed to be a separate tool or a python plugin with functionality accessible to any website by its owner?

**Comment by HelloZeroNet** [comment]
It should be a simple zeronet site

**Comment by obv-mikhail** [comment]
Instead of using configuration files we can use a simpler component approach.

/js/utils/AdminInterface.coffee
`
class AdminInterface extends ZeroFrame
  constructor: (title) ->
    @title = title
		@pages = []
		
  render: () => 
    h("div.AdminInterface",
      {title: @title}, 
      pages.map => Page
    )

  Page: (obj) => 
    h("div.Page", 
      {name: obj.name, title: obj.title, orderby: obj.orderby}, 
      obj.inside.map => type
    )
	
  type: (obj) =>
    if obj.type == "user"
      User(obj)

  User: (obj) =>

  addPage: (name, title, orderby) =>
    pages.push {name: name, title: title, orderby: orderby, inside: []}

  addType: (index) =>
    pages.index.inside.push #stuff
`

/js/utils/ZeroTalkAdmin.coffee
`
class ZeroTalkAdmin extends ZeroFrame
  init: => 
    panel = AdminInterface("ZeroTalkAdmin")
    panel.addPage "Topic", "Topics", "added DESC"
    panel.addType 0 #stuff
    projector.append document.body, panel.render
`

**Comment by HelloZeroNet** [comment]
pure js also could work, have to try which one is easier to read/extend by creating examples for zerotalk/zeroblog/zeromail

**Comment by obv-mikhail** [comment]
This how it would look like for ZeroTalk.
```Coffescript
class AdminInterface extends ZeroFrame
  constructor: (title, pages) ->
    @title = title
    @pages = pages
    @data = ""
		
  render: => 

  page: (name, attr, cont) -> h(name+".page", attr, cont)

  user: (name, attr) -> h(name+".user", attr)
  input: (name, attr) -> h(name+".input", attr)
  textarea: (name, attr) -> h(name+".textarea", attr)
  select: (name, attr) -> h(name+".select", attr)

class ZeroTalkAdmin extends ZeroFrame
  init: -> 
    ai = AdminInterface
    panel = ai("ZeroTalkAdmin", [
      ai.Page("Topic", {title:"Topics", orderby: "added DESC"}, [
        ai.user("json_id", {title: "Username", disabled: 1, save: 0}),
        ai.input("added", {formatter: Date.toDate, deformatter: Date.fromDate}),
        ai.input("title"),
        ai.textarea("body"),
        ai.select("type", {values: ["", "group"]}),
        ai.select("parent_topic_uri", {title: "Parent topic", values: Topics.listUris})
        ]),
      ai.Page("comment", {title:"Comments", orderby: "added DESC", parent: topic}, [
        ai.user("json_id", {title: "Username", disabled: 1, save: 0}),
        ai.input("added", {formatter: Date.toDate, deformatter: Date.fromDate})
        ai.textarea("body")
        ai.select("type", {values: ["", "group"]})
        ai.select("topic_uri", {title: "Topic", values: Topics.listUris})
      ]),
      ai.Page("settings", {title:"Settings", file: "content.json"}, [
        ai.input("settings.admin", {title: "Admin name"})
        ai.input("settings.href", {title: "Admin contact url"})
        ai.select("settings.sticky_uris", {title: "Sticky topic uris", multi: 1, values: Topics.listUris})
      ])
    ]
    projector.append document.body, panel.render
```

**Comment by HelloZeroNet** [comment]
looks good for me

**Comment by Snarkly** [comment]
Admin Panel tutorials https://www.youtube.com/watch?v=oxGmXOKazw0 https://www.youtube.com/watch?v=SSAQK4Nwkm4&feature=youtu.be https://www.youtube.com/watch?v=mysAB7NAjDk&feature=youtu.be https://www.youtube.com/watch?v=Ct8NKKlsiMM

**Comment by obv-mikhail** [comment]
How does dbQuery query work? I am currently having to use timeouts to make sure that data is fetched before proceeding with rendering. Is there a different workaround for this? 

Can the database file be loaded into the site with fileGet? If so it might be a good solution to use sql.js.

**Comment by HelloZeroNet** [comment]
Its verry simple, but the chat tutorial has example for dbquery. Loading database file using fileGet is a bad idea and does not makes much sense.

**Comment by obv-mikhail** [comment]
How would I access 'table' variable from within the callback function?

```coffeescript
		for table in @tables
			Page.cmd "dbQuery", ["SELECT * FROM "+table.name], (res) =>
				@columns[table.name] = res
```

**Comment by HelloZeroNet** [comment]
This should work:
```
checkTable: (table_name) =>
    Page.cmd "dbQuery", ["SELECT * FROM "+table_name], (res) =>
                @columns[table_name] = res

for table in @tables
   checkTable(table.name)
```
part to separate function, then it should work

**Comment by obv-mikhail** [comment]
https://github.com/obv-mikhail/ZeroNet-Admin-Interface/
So far able to retrieve data based on provided configuration. Next will work on storing data back to the database.

Live demo here: http://127.0.0.1:43110/1DWD4M2dy6rDAF1dhxQE953hSWTib1ZEjo/ 

If no seeds, here is a picture:
http://i.imgur.com/d66TF6g.png?1

**Comment by obv-mikhail** [comment]
A site database can be updated with fileWrite, correct? Is using userPublickey to get inner path the correct approach? If so, how can userPublickey be called?

The below code doesn't work.
`Page.cmd userPublickey ["1"]`

**Comment by HelloZeroNet** [comment]
userPublickey is for ecies encryption. the user's auth_address is in site_info object (please check the zerochat tutorial)

**Comment by obv-mikhail** [comment]
To update data.json, should "added" always be treated as the unique identifier for any configurations? "topic_id" would also work, but it wasn't specified in the configuration for ZeroTalk.

**Comment by Snarkly** [comment]
How is this project coming along?

**Comment by Snarkly** [comment]
I found a few coders on youtube who know about admin panels and I'm contacting them to see if they can help create an admin panel for ZeroNet.

**Comment by obv-mikhail** [comment]
HelloZeroNet, maybe create an official (empty for now) repository, like for ZeroTalk etc? Would make more sense to discuss things in there.

**Comment by HelloZeroNet** [comment]
Good idea, created and added some thing before start coding: https://github.com/HelloZeroNet/ZeroAdmin/issues
